lib/utils/compose.js

```diff
@@ -0,0 +1,1087 @@
+		if (end == null) {
+			end = false;
+		}
+		const moment = require('moment');
+		require('moment-duration-format')(moment);
+
+		this._tty.clearLine();
+		this._tty.write(this._prefix);
+		if (end && this._cancelled) {
+			this._tty.writeLine('Build cancelled');
+		} else if (end) {
+			const serviceCount = this._services.length;
+			const serviceStr =
+				serviceCount === 1 ? '1 service' : `${serviceCount} services`;
+			const durationStr = this._startTime == null ? 'unknown time' : moment
+				.duration(Math.floor((Date.now() - this._startTime) / 1000), 'seconds')
+				.format();
+			this._tty.writeLine(`Built ${serviceStr} in ${durationStr}`);
+		} else {
+			this._tty.writeLine(`Building services... ${this._spinner()}`);
+		}
+	}
+
+	_renderSummary(serviceToStrMap) {
+		const _ = require('lodash');
+		const chalk = getChalk();
+		const truncate = require('cli-truncate');
+		const strlen = require('string-width');
+
+		this._services.forEach((service, index) => {
+			let str = _.padEnd(this._prefix + chalk.bold(service), this._prefixWidth);
+			str += serviceToStrMap[service];
+			if (this._maxLineWidth != null) {
+				str = truncate(str, this._maxLineWidth);
+			}
+			this._lineWidths[index] = strlen(str);
+
+			this._tty.clearLine();
+			this._tty.writeLine(str);
+		});
+	}
+}
+
+class BuildProgressInline {
+	constructor(outStream, descriptors) {
+		this.start = this.start.bind(this);
+		this.end = this.end.bind(this);
+		this._renderEvent = this._renderEvent.bind(this);
+		const _ = require('lodash');
+		const through = require('through2');
+
+		const services = _.map(descriptors, 'serviceName');
+		const eventHandler = this._renderEvent;
+		const streams = _(services)
+			.map(function(service) {
+				const stream = through.obj(function(event, _enc, cb) {
+					eventHandler(service, event);
+					return cb();
+				});
+				stream.pipe(outStream, { end: false });
+				return [service, stream];
+			})
+			.fromPairs()
+			.value();
+
+		const offset = 10; // account for escape sequences inserted for colouring
+		this._prefixWidth = offset + _.max(_.map(services, 'length'));
+		this._outStream = outStream;
+		this._services = services;
+		this._startTime = null;
+		this._ended = false;
+
+		this.streams = streams;
+	}
+
+	start() {
+		this._outStream.write('Building services...\n');
+		this._services.forEach(service => {
+			this.streams[service].write({ status: 'Preparing...' });
+		});
+		this._startTime = Date.now();
+	}
+
+	end(summary = null) {
+		const moment = require('moment');
+		require('moment-duration-format')(moment);
+
+		if (this._ended) {
+			return;
+		}
+		this._ended = true;
+
+		if (summary != null) {
+			this._services.forEach(service => {
+				this._renderEvent(service, summary[service]);
+			});
+		}
+
+		const serviceCount = this._services.length;
+		const serviceStr =
+			serviceCount === 1 ? '1 service' : `${serviceCount} services`;
+		const durationStr = this._startTime == null ? 'unknown time' : moment
```

Comment:
This was changed to have the `this._startTime == null ? 'unknown time' :` check since the type for `this._startTime` is nullable